### PR TITLE
[Verifier] Support reuse of source query output table

### PR DESCRIPTION
**Pull request by gggrace14**
Add support to Presto Verifier to reuse the output tables of the source query for control and test. Add configs --control-reuse-table and --test-reuse-table to turn the feature on and off. If reuse-table is turned on, skip the execution of main query as well as the setup and teardown queries. Compute checksums on the output table of main query directly.

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Comment by mbasmanova** [comment]
@gggrace14 Ge, how much faster verifier runs with this optimization?

**Comment by spershin** [comment]
> @gggrace14 Ge, how much faster verifier runs with this optimization?

@mbasmanova 

This needs to be thoroughly tested to get some numbers and also depends on workflow and number of queries.
We can run this for a while in Native and compare the difference.

But, obviously it is the time without optimization minus time it takes to run setup, main and teardown queries.
Note, that it looks like (I noticed) the verifier runs control query 1st and then runs the test one, so both main queries are SERIALIZED.
Not sure if there is a verifier option to run them in parallel.

**Comment by gggrace14** [comment]
class QueryRewriter, class AbstractVerification and class VerifierDao have the main changes. Others are about passing the required info around.

File: presto-verifier/src/main/java/com/facebook/presto/verifier/source/VerifierDao.java
Comment (spershin) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/source/VerifierDao.java:38]: Strange, didn't we put control and test query Ids into the table before this change?
The ones that you click on, like "Control Query ID".

**Comment by gggrace14** [comment]
> Not sure if there is a verifier option to run them in parallel.

For control & test checksum queries, current implementation is to run them in sequence. For main queries, param --concurrent-control-and-test can run them in parallel.

File: presto-verifier/pom.xml
Comment (spershin) [inline presto-verifier/pom.xml:52]: What do we need from the presto-client in the verifier?

**Review by spershin — APPROVED** [review]
@gggrace14
Thanks for this change (and others) to enabled re-use of the tables and reduce verification time and compute.

The change looks good to me, few questions.

Maybe worth @kewang1024, @ajaygeorge  or @aweisberg to take a quick look at it too.

File: presto-verifier/src/main/java/com/facebook/presto/verifier/source/VerifierDao.java
Comment (gggrace14) [reply presto-verifier/src/main/java/com/facebook/presto/verifier/source/VerifierDao.java:38]: Yeah, the "Control Query ID" link on UI is populated with the query_id returned by the actual control query run triggered by Verifier today. If we reuse the shadow/prod query output and don't trigger the control/test query runs, we have to pass the query_ids to Verifer through xdb, so that the UI can display them.

Otherwise our debugging experience is bad. Found it out from testing.

File: presto-verifier/pom.xml
Comment (gggrace14) [reply presto-verifier/pom.xml:52]: One place is the following lines of change. The change is about passing the query_id from xdb to the UI by setting them in the related data structures.

In presto-verifier
```
import com.facebook.presto.client.StatementStats;
...
public static QueryActionStats queryIdStats(String queryId)
    {
        return new QueryActionStats(Optional.of(QueryStats.create(queryId, StatementStats.builder().setState(QueryState.FINISHED.toString()).build())), Optional.empty());
    }
```

File: presto-verifier/src/main/java/com/facebook/presto/verifier/framework/AbstractVerification.java
Comment (kewang1024) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/framework/AbstractVerification.java:212]: It would be clearer to put the logic inside the getQueryRewrite()

File: presto-verifier/src/main/java/com/facebook/presto/verifier/framework/AbstractVerification.java
Comment (kewang1024) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/framework/AbstractVerification.java:220]: same as above

File: presto-verifier/src/main/java/com/facebook/presto/verifier/rewrite/QueryRewriteConfig.java
Comment (kewang1024) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/rewrite/QueryRewriteConfig.java:41]: 1.Is it the best effort?
if it is set to true, but the output table is deleted, does the verification fail or follow original routine?

2.Does it only apply to INSERT and CTAS queries? if so, we should specify in the config description

**Comment by steveburnett** [comment]
Suggest adding 

* documentation for the new configuration properties. 

  A good location might be near the `control.table-prefix` and `test.table-prefix` entries (lines 275-276) in https://github.com/prestodb/presto/blob/master/presto-docs/src/main/sphinx/admin/verifier.rst?plain=1. 

* a release note entry. Perhaps something like the following:

```
== RELEASE NOTES ==

Verifier Changes
* Add  `control-reuse-table` and `test-reuse-table` configuration properties for the Presto Verifier to reuse the output tables of the source query for control and test :pr:`22169`
```

File: presto-verifier/src/test/java/com/facebook/presto/verifier/rewrite/TestQueryRewriter.java
Comment (arhimondr) [inline presto-verifier/src/test/java/com/facebook/presto/verifier/rewrite/TestQueryRewriter.java:487]: nit: new line

File: presto-verifier/src/main/java/com/facebook/presto/verifier/rewrite/QueryRewriter.java
Comment (arhimondr) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/rewrite/QueryRewriter.java:182]: What if the original table contains some other data then the data inserted by the query being verifier. How do we know what partitions of the original table to compute the checksum over?

File: presto-verifier/src/main/java/com/facebook/presto/verifier/framework/AbstractVerification.java
Comment (arhimondr) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/framework/AbstractVerification.java:323]: nit: maybe wrap `testQueryContext.getState() != QueryState.SUCCEEDED && testQueryContext.getState() != QueryState.REUSE` into `( .. )`

File: presto-verifier/src/main/java/com/facebook/presto/verifier/framework/AbstractVerification.java
Comment (arhimondr) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/framework/AbstractVerification.java:82]: nit: consider keeping it final and return a new instance in `VerificationResult` (instead of `this`)

File: presto-verifier/src/main/java/com/facebook/presto/verifier/event/QueryInfo.java
Comment (gggrace14) [reply presto-verifier/src/main/java/com/facebook/presto/verifier/event/QueryInfo.java:56]: Ah, typo

File: presto-verifier/src/main/java/com/facebook/presto/verifier/rewrite/QueryRewriter.java
Comment (gggrace14) [reply presto-verifier/src/main/java/com/facebook/presto/verifier/rewrite/QueryRewriter.java:182]: A good question! I have a next PR on top of this one that assembles a partition predicate and applies it to the checksum query.

File: presto-verifier/src/main/java/com/facebook/presto/verifier/framework/AbstractVerification.java
Comment (gggrace14) [reply presto-verifier/src/main/java/com/facebook/presto/verifier/framework/AbstractVerification.java:323]: Revised

File: presto-verifier/src/main/java/com/facebook/presto/verifier/framework/VerificationManager.java
Comment (gggrace14) [reply presto-verifier/src/main/java/com/facebook/presto/verifier/framework/VerificationManager.java:356]: !event.isPresent() is when the test case is to be resubmitted. From our experience, we find it confusing if we also add the resubmission count to the counter SKIPPED. This problem becomes more apparent when reuse-table is enabled, as we resubmit more to be safe.

The main condition of SKIPPED (line 361) is when the control query run fails, ex., due to syntax error, which is how ppl usually assume the counter SKIPPED. When SKIPPED includes resubmissions as today, we find it confusing that the counters SUCCEED, FAILED and SKIPPED don't add up to the total test cases, as one test case can be resubmitted multiple times. So I split it in this change and add one more counter RESUBMITTED (line 366).

File: presto-verifier/src/main/java/com/facebook/presto/verifier/source/PrestoQuerySourceQuerySupplier.java
Comment (gggrace14) [reply presto-verifier/src/main/java/com/facebook/presto/verifier/source/PrestoQuerySourceQuerySupplier.java:44]: Revised here.

File: presto-verifier/src/main/java/com/facebook/presto/verifier/framework/AbstractVerification.java
Comment (gggrace14) [reply presto-verifier/src/main/java/com/facebook/presto/verifier/framework/AbstractVerification.java:82]: A good catch. Also thought it's better if I can make this class immutable. However, I can't figure out how the constructor can look like, esp. since this class AbstractVerification is an abstract class. Please let me know if you have a good suggestion.

I thought I could add an abstract method to this class
```
public abstract AbstractVerification withSourceQuery(SourceQuery sourceQuery);
```
Then its subclass DataVerification overrides this method, creates and returns a DataVerification instance. However, we can't access those private fields, ex. queryActions, of the base abstract class in the subclass when calling the DataVerification constructor.

File: presto-jdbc/pom.xml
Comment (gggrace14) [reply presto-jdbc/pom.xml]: Figured out a way to avoid this attribute and changing the packaging.

**Comment by gggrace14** [comment]
> Suggest adding
> 
> * documentation for the new configuration properties.
>   A good location might be near the `control.table-prefix` and `test.table-prefix` entries (lines 275-276) in https://github.com/prestodb/presto/blob/master/presto-docs/src/main/sphinx/admin/verifier.rst?plain=1.
> * a release note entry. Perhaps something like the following:
> 
> ```
> == RELEASE NOTES ==
> 
> Verifier Changes
> * Add  `control-reuse-table` and `test-reuse-table` configuration properties for the Presto Verifier to reuse the output tables of the source query for control and test :pr:`22169`
> ```

Adding the corresponding documentation to the verifier.rst.

For the release note, I have a next PR <link> on top of this for the complete feature. I'd prefer to add release note when the complete feature is in, to expose it to the user.

File: presto-verifier/src/main/java/com/facebook/presto/verifier/rewrite/QueryRewriteConfig.java
Comment (gggrace14) [reply presto-verifier/src/main/java/com/facebook/presto/verifier/rewrite/QueryRewriteConfig.java:41]: 1. If the output table is deleted, the verification will fail.
2. Revised.

File: presto-docs/src/main/sphinx/admin/verifier.rst
Comment (steveburnett) [inline presto-docs/src/main/sphinx/admin/verifier.rst:279]:
```suggestion
``control.reuse-table``                     If ``true``, reuse the output table of the control source Insert and CreateTableAsSelect
```
Nit for formatting, same as #22965.

File: presto-docs/src/main/sphinx/admin/verifier.rst
Comment (steveburnett) [inline presto-docs/src/main/sphinx/admin/verifier.rst:281]:
```suggestion
``test.reuse-table``                        If ``true``, reuse the output table of the test source Insert and CreateTableAsSelect
```

**Review by steveburnett — CHANGES_REQUESTED** [review]
Thanks for the doc! Nit of formatting, no other concerns.

File: presto-docs/src/main/sphinx/admin/verifier.rst
Comment (gggrace14) [reply presto-docs/src/main/sphinx/admin/verifier.rst:279]: Revised

File: presto-docs/src/main/sphinx/admin/verifier.rst
Comment (gggrace14) [reply presto-docs/src/main/sphinx/admin/verifier.rst:281]: Revised

File: presto-verifier/src/main/java/com/facebook/presto/verifier/framework/AbstractVerification.java
Comment (gggrace14) [reply presto-verifier/src/main/java/com/facebook/presto/verifier/framework/AbstractVerification.java]: Hi @arhimondr , I figured out a different way that doesn't need to make this sourceQuery field mutable. The original objective is to pass a signal so that the "resubmitted" verification workflow will not reuse output tables. Previously I constructed a new SourceQuery instance. A different way I figured is to check the `VerificationContext.getResubmissionCount()` as below. I figured I'm able to access VerificationContext when calling `queryRewriter.rewriteQuery()`
```
    @Override
    protected QueryObjectBundle getQueryRewrite(ClusterType clusterType)
    {
        return queryRewriter.rewriteQuery(getSourceQuery().getQuery(clusterType), getSourceQuery().getQueryConfiguration(clusterType), clusterType,
                getVerificationContext().getResubmissionCount() == 0);
    }
```

**Review by arhimondr — APPROVED** [review]

**Comment by gggrace14** [comment]
> Thanks for the doc! Nit of formatting, no other concerns.

Hi @steveburnett , revised the format of verifier.rst document. Would you be able to help review again?

**Comment by gggrace14** [comment]
> > Thanks for the doc! Nit of formatting, no other concerns.
> 
> Hi @steveburnett , revised the format of verifier.rst document. Would you be able to help review again?

Hi @steveburnett , thanks for the review! It looks to need your approval to be able to merge.

**Comment by steveburnett** [comment]
Not a required test so not blocking merge, but the test 

`singlestore tests / singlestore-dockerized-tests (pull_request)` 

failed here. I believe I've seen other mentions of that test being flaky lately, wanted to mention it in case people are looking for examples of it failing.

**Comment by gggrace14** [comment]
> LGTM! (docs)
> 
> Pull updated branch, new local build, looks good. Thanks!

Thank you, @steveburnett !

**Pull request merged**
